piptools/build.py

```diff
@@ -0,0 +1,159 @@
+from importlib import metadata as importlib_metadata
+from typing import Any, Iterator
+
+import build
+import build.env
+import pyproject_hooks
+from pip._internal.req import InstallRequirement
+from pip._internal.req.constructors import install_req_from_line, parse_req_from_line
+
+PYPROJECT_TOML = "pyproject.toml"
+
+
+@dataclass
+class ProjectMetadata:
+    extras: tuple[str, ...]
+    requirements: tuple[InstallRequirement, ...]
+    build_requirements: tuple[InstallRequirement, ...]
+
+
+def build_project_metadata(
+    src_file: str,
+    build_distributions: tuple[str, ...],
+    *,
+    isolated: bool,
+    quiet: bool,
+) -> ProjectMetadata:
+    """
+    Return the metadata for a project.
+
+    Uses the ``prepare_metadata_for_build_wheel`` hook for the wheel metadata
+    if available, otherwise ``build_wheel``.
+
+    Uses the ``prepare_metadata_for_build_{dist}`` hook for each ``build_distributions``
+    if available.
+
+    :param src_file: Project source file
+    :param build_distributions: A tuple of build distributions to get the dependencies
+                                of (``sdist`` or ``wheel`` or ``editable``).
+    :param isolated: Whether to run invoke the backend in the current
+                     environment or to create an isolated one and invoke it
+                     there.
+    :param quiet: Whether to suppress the output of subprocesses.
+    """
+
+    src_dir = os.path.dirname(os.path.abspath(src_file))
+    with _create_project_builder(src_dir, isolated=isolated, quiet=quiet) as builder:
+        metadata = _build_project_wheel_metadata(builder)
+        extras = tuple(metadata.get_all("Provides-Extra") or ())
+        requirements = tuple(
+            _prepare_requirements(metadata=metadata, src_file=src_file)
+        )
+        build_requirements = tuple(
+            _prepare_build_requirements(
+                builder=builder,
+                distributions=build_distributions,
+                package_name=_get_name(metadata),
+            )
+        )
+        return ProjectMetadata(
+            extras=extras,
+            requirements=requirements,
+            build_requirements=build_requirements,
+        )
+
+
+@contextlib.contextmanager
+def _create_project_builder(
+    src_dir: str, *, isolated: bool, quiet: bool
+) -> Iterator[build.ProjectBuilder]:
+    builder = build.ProjectBuilder(
+        os.fspath(src_dir),
+        runner=pyproject_hooks.quiet_subprocess_runner
+        if quiet
+        else pyproject_hooks.default_subprocess_runner,
+    )
+
+    if not isolated:
+        yield builder
+        return
+
+    with build.env.IsolatedEnvBuilder() as env:
+        builder.python_executable = env.executable
+        builder.scripts_dir = env.scripts_dir
+        env.install(builder.build_system_requires)
+        env.install(builder.get_requires_for_build("wheel"))
+        yield builder
+
+
+def _build_project_wheel_metadata(
+    builder: build.ProjectBuilder,
+) -> Any:
```

Comment:
Maybe `cast(importlib_metadata.PackageMetadata)` in return instead of `Any`?

Reply:
I couldn't get type hinting with `PackageMetadata` to work, which is why I went with `Message`. This particular proposal results in `piptools/build.py:103: error: Name "importlib_metadata.PackageMetadata" is not defined  [name-defined]`

Reply:
That is probably version-related + `importlib_metadata` will be newer than it's `importlib.metadata` stdlib counterpart most of the time...

Version-based type selection would likely address this.

Reply:
As far as I can tell `importlib_metadata` is an alias for `importlib.metadata`, presumably to confusion with variables called `metadata`. But you are right in that it is version dependent. I spent some more time with this now and I think I figured out something that will work and naturally shed entropy as we step the python version.

Reply:
Correct.